op-service/txmgr/txmgr.go

```diff
@@ -429,12 +417,45 @@ func (m *SimpleTxManager) craftTx(ctx context.Context, candidate TxCandidate) (*
 			GasFeeCap: gasFeeCap,
 			Value:     candidate.Value,
 			Data:      candidate.TxData,
-			Gas:       gasLimit,
+			Gas:       candidate.GasLimit,
 		}
 	}
 	return m.signWithNextNonce(ctx, txMessage) // signer sets the nonce field of the tx
 }
 
+// estimateOrValidateCandidateTxGas estimates the gas limit if it's not set, or validates it by calling CallContract.
+// It returns the gas limit (which may be updated if it was estimated) and an error.
```

Comment:
```suggestion
// estimateOrValidateCandidateTxGas either:
// a) validates and returns the candidate.GasLimit (if set) using CallContract
// b) estimates the gas limit using backend.EstimatGas and returns it. 
```
I don't think the candidate is ever mutated by this function, which seems to be implied when I read the godoc?